### DISCUSSION TITLE
How to trigger a single build on either push or pull_request

### DISCUSSION THREAD

**Discussion by gitfool** [community/community#26872] [Actions]
<p>I’m already using Azure Pipelines for builds. I’m now trying GitHub Actions for builds side by side, so I can compare them.</p>
<p>I subsequently received a PR for one of my projects from dependabot (<a href="https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53" rel="nofollow noopener">https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53</a>). Azure Pipelines only triggered one build, as usual and as expected, but GitHub Actions triggered two builds; one for push and one for pull_request:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0322c4548b2b476a51e72258cfcd834ad8b6aeb6.jpeg" data-download-href="/uploads/short-url/rJV3brddz7jd8pAS7j7Zsuariu.jpg?dl=1" title="Checks.jpg" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/750121/181100401-cfc93884-68bd-4f0a-90b8-e8ac0774dbf9.jpg" alt="Checks.jpg" data-base62-sha1="rJV3brddz7jd8pAS7j7Zsuariu" width="690" height="317" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/0/0322c4548b2b476a51e72258cfcd834ad8b6aeb6_2_690x317.jpg, https://aws1.discourse-cdn.com/github/original/2X/0/0322c4548b2b476a51e72258cfcd834ad8b6aeb6.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/0/0322c4548b2b476a51e72258cfcd834ad8b6aeb6.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0322c4548b2b476a51e72258cfcd834ad8b6aeb6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Checks.jpg</span><span class="informations">739×340 96.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Is this a bug or do I need to configure the GitHub Actions differently to match Azure Pipelines behaviour? (After merging the pull request there should again be just a single build, but which should not be due to a pull request.)</p>
<p><a href="https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.azure-pipelines.yml" rel="nofollow noopener">https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.azure-pipelines.yml</a></p>
<p><a href="https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.github/workflows/build.yml" rel="nofollow noopener">https://github.com/gitfool/BoardGameGeek.Dungeon/blob/master/.github/workflows/build.yml</a></p>

**Comment by paulineribeyre** [comment]
<p>Hello, your workflow is configured with “on: [push, pull_request]” so it gets triggered twice when a commit is pushed to open a pull request. It looks to me like “on: push” would work well for your use case.</p>

**Comment by gitfool** [comment]
<p>The devil in the details is that pull requests should not publish build artifacts, whereas pushes / merges to master should, so I need to be able to differentiate pull requests from pushes with the GITHUB environment variables:</p>
<p><a href="https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53/checks?check_run_id=228755652#step:6:34" rel="nofollow noopener">https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53/checks?check_run_id=228755652#step:6:34</a></p>
<p><a href="https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53/checks?check_run_id=228755704#step:6:34" rel="nofollow noopener">https://github.com/gitfool/BoardGameGeek.Dungeon/pull/53/checks?check_run_id=228755704#step:6:34</a></p>
<p>Azure Pipelines makes this easy and obvious:</p>
<p><a href="https://dev.azure.com/gitfool/BoardGameGeek.Dungeon/_build/results?buildId=304&amp;view=logs&amp;jobId=1c8ce01e-97de-5be9-a863-648419f09741&amp;taskId=09181c4b-3f3c-50e9-b57f-22616eaa56ea&amp;lineStart=134&amp;lineEnd=142&amp;colStart=1&amp;colEnd=1" rel="nofollow noopener">https://dev.azure.com/gitfool/BoardGameGeek.Dungeon/_build/results?buildId=304&amp;view=logs&amp;jobId=1c8ce01e-97de-5be9-a863-648419f09741&amp;taskId=09181c4b-3f3c-50e9-b57f-22616eaa56ea&amp;lineStart=134&amp;lineEnd=142&amp;colStart=1&amp;colEnd=1</a></p>

**Comment by ethomson** [comment, marked as answer]
<p>You can specify the branches that you want builds to run on.  WIth the <code>on: [push, pull_request]</code> setting, you’ll get any pushes into _any_ branch, and pull requests into _any_ branch.</p>
<p>This means that when dependabot creates its branch - in this case,</p>
<p><code>dependabot/nuget/Microsoft.NET.Test.Sdk-16.3.0</code> - that your build will run.  It will then run again when it creates the pull request from that branch.</p>
<p>If you only want to run when something has a pull request opened into master, or when something is merged into master, then you can specify this as:</p>
<pre><code>on:
  push:
    branches:
    - master
  pull_request:
    branches:
    - master
</code></pre>

**Comment by fkirc** [comment]
<p>I published <a href="https://github.com/marketplace/actions/skip-duplicate-actions" rel="noopener nofollow ugc">https://github.com/marketplace/actions/skip-duplicate-actions</a> to solve this problem in an easy way.<br>
You can give it a try if this is still an issue.</p>
